hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java

```diff
@@ -215,253 +305,265 @@ public static void addDebugLevel(Integer level) {
-                      LOG.warn(e.getMessage());
-                    }
-                    infileNode = new CompactionNode(infile,
-                        lastSnapshotPrefix, numKeys,
-                        UNKNOWN_COMPACTION_GEN);
-                    compactionDAGFwd.addNode(infileNode);
-                    compactionDAGReverse.addNode(infileNode);
-                    compactionNodeTable.put(infile, infileNode);
-                  }
-                  if (outfileNode.fileName.compareToIgnoreCase(
-                      infileNode.fileName) != 0) {
-                    compactionDAGFwd.putEdge(outfileNode, infileNode);
-                    compactionDAGReverse.putEdge(infileNode, outfileNode);
-                  }
-                }
-              }
-              if (debugEnabled(DEBUG_DAG_BUILD_UP)) {
-                printMutableGraph(null, null,
-                    compactionDAGFwd);
-              }
+    setRocksDBForCompactionTracking(rocksOptions, new ArrayList<>());
+  }
+
+  private AbstractEventListener newCompactionBeginListener() {
+    return new AbstractEventListener() {
+      @Override
+      public void onCompactionBegin(RocksDB db,
+          CompactionJobInfo compactionJobInfo) {
+
+        synchronized (compactionListenerWriteLock) {
+
+          if (compactionJobInfo.inputFiles().size() == 0) {
+            LOG.error("Compaction input files list is empty");
+            return;
+          }
+
+          // Create hardlink backups for the SST files that are going
+          // to be deleted after this RDB compaction.
+          for (String file : compactionJobInfo.inputFiles()) {
+            LOG.debug("Creating hard link for '{}'", file);
+            String saveLinkFileName =
+                sstBackupDir + new File(file).getName();
+            Path link = Paths.get(saveLinkFileName);
+            Path srcFile = Paths.get(file);
+            try {
+              Files.createLink(link, srcFile);
+            } catch (IOException e) {
+              LOG.error("Exception in creating hard link for {}", file);
+              throw new RuntimeException("Failed to create hard link", e);
             }
           }
-        };
 
-    list.add(onCompactionCompletedListener);
-    rocksOptions.setListeners(list);
+        }
+      }
+    };
   }
 
-  public RocksDB getRocksDBInstanceWithCompactionTracking(String dbPath)
-      throws RocksDBException {
-    final Options opt = new Options().setCreateIfMissing(true)
-        .setCompressionType(CompressionType.NO_COMPRESSION);
-    opt.setMaxBytesForLevelMultiplier(2);
-    setRocksDBForCompactionTracking(opt);
-    return RocksDB.open(opt, dbPath);
+  /**
+   * Helper function to append the list of SST files to a StringBuilder
+   * for a compaction log entry. Does not append a new line.
+   */
+  private static void appendCompactionLogStringBuilder(List<String> files,
+      StringBuilder sb) {
+
+    Iterator<String> it = files.iterator();
+    while (it.hasNext()) {
```

Comment:
nit: You can use `String.join(",", files);` or `files.stream().collect(Collectors.joining(","));`.

Reply:
Good one! Done